api/clients/eigenda_client_v2.go

```diff
@@ -0,0 +1,142 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/api/clients/codecs"
+	corev2 "github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/hashicorp/go-multierror"
+	"math/rand"
+)
+
+// EigenDAClientV2 provides the ability to get blobs from the relay subsystem, and to send new blobs to the disperser.
+type EigenDAClientV2 interface {
+	// GetBlob iteratively attempts to retrieve a given blob with key blobKey from the relays listed in the blobCertificate.
+	GetBlob(ctx context.Context, blobKey corev2.BlobKey, blobCertificate corev2.BlobCertificate) ([]byte, error)
+	GetCodec() codecs.BlobCodec
+	Close() error
+}
+
+type eigenDAClientV2 struct {
+	clientConfig EigenDAClientConfig
+	log          logging.Logger
+	relayClient  RelayClient
+	codec        codecs.BlobCodec
+}
+
+var _ EigenDAClientV2 = &eigenDAClientV2{}
+
+// BuildEigenDAClientV2 builds an EigenDAClientV2 from config structs.
+func BuildEigenDAClientV2(
+	log logging.Logger,
+	clientConfig EigenDAClientConfig,
+	relayClientConfig RelayClientConfig) (EigenDAClientV2, error) {
+
+	err := clientConfig.CheckAndSetDefaults()
+	if err != nil {
+		return nil, err
+	}
+
+	relayClient, err := NewRelayClient(&relayClientConfig, log)
+
+	lowLevelCodec, err := codecs.BlobEncodingVersionToCodec(clientConfig.PutBlobEncodingVersion)
+	if err != nil {
+		return nil, fmt.Errorf("create low level codec: %w", err)
+	}
+	var codec codecs.BlobCodec
+	if clientConfig.DisablePointVerificationMode {
+		codec = codecs.NewNoIFFTCodec(lowLevelCodec)
+	} else {
+		codec = codecs.NewIFFTCodec(lowLevelCodec)
+	}
+
+	return NewEigenDAClientV2(log, clientConfig, relayClient, codec)
+}
+
+// NewEigenDAClientV2 assembles an EigenDAClientV2 from subcomponents that have already been constructed and initialized.
+func NewEigenDAClientV2(
+	log logging.Logger,
+	clientConfig EigenDAClientConfig,
+	relayClient RelayClient,
+	codec codecs.BlobCodec) (EigenDAClientV2, error) {
+
+	return &eigenDAClientV2{
+		clientConfig: clientConfig,
+		log:          log,
+		relayClient:  relayClient,
+		codec:        codec,
+	}, nil
+}
+
+// GetBlob iteratively attempts to retrieve a given blob with key blobKey from the relays listed in the blobCertificate.
+//
+// The relays are attempted in random order.
+//
+// The returned blob is decoded.
+func (c *eigenDAClientV2) GetBlob(ctx context.Context, blobKey corev2.BlobKey, blobCertificate corev2.BlobCertificate) ([]byte, error) {
+	// create a randomized array of indices, so that it isn't always the first relay in the list which gets hit
+	random := rand.New(rand.NewSource(rand.Int63()))
```

Comment:
if we want to go with this random appraoch (but see my other comment), maybe we should pass the source of randomness in the constructor and reuse it everywhere? To make tests deterministic.

Reply:
Passed in a source of randomness for now, pending resolution of [this discussion](https://github.com/Layr-Labs/eigenda/pull/972#discussion_r1878297938)